sdk-core/src/main/java/care/data4life/sdk/CryptoSecureStore.kt

```diff
@@ -0,0 +1,102 @@
+ */
+package care.data4life.sdk
+
+import care.data4life.crypto.ExchangeKey
+import care.data4life.crypto.GCKey
+import care.data4life.crypto.GCKeyPair
+import care.data4life.crypto.KeyType
+import care.data4life.crypto.error.CryptoException
+import care.data4life.sdk.crypto.CryptoContract
+import care.data4life.sdk.lang.D4LException
+import care.data4life.securestore.SecureStoreContract
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import java.io.IOException
+import java.lang.reflect.Type
+
+class CryptoSecureStore @JvmOverloads constructor(
+    moshi: Moshi? = null,
+    private val secureStore: SecureStoreContract.SecureStore
+) : CryptoContract.CryptoSecureStore {
+    private val moshi: Moshi = if (moshi is Moshi) moshi else Moshi.Builder().build()
+
+    override fun clear(): Unit = secureStore.clear()
+
+    override fun storeSecret(
+        alias: String,
+        secret: CharArray
+    ): Unit = secureStore.addData(alias, secret)
+
+    @Throws(D4LException::class)
+    override fun getSecret(alias: String): CharArray {
+        return secureStore.getData(alias)
+            ?: throw CryptoException.DecryptionFailed("Failed to decrypt data")
+    }
+
+    override fun <T : Any> storeSecret(alias: String, crate: T) {
+        val type: Type = Types.getRawType(crate::class.java)
+        val jsonData = moshi.adapter<Any>(type).toJson(crate)
+        storeSecret(alias, jsonData.toCharArray())
+    }
+
+    @Throws(D4LException::class)
+    override fun <T> getSecret(alias: String, type: Class<T>): T? {
+        val data: CharArray = secureStore.getData(alias)
+            ?: throw CryptoException.DecryptionFailed("Failed to decrypt data")
+
+        return try {
+            moshi.adapter(type).fromJson(String(data))
+        } catch (e: IOException) {
+            throw CryptoException.DecryptionFailed("Failed to decrypt data")
+        }
+    }
+
+    override fun deleteSecret(alias: String): Unit = secureStore.removeData(alias)
+
+    override fun storeKey(alias: String, key: GCKey, keyType: KeyType) {
+        val exchangeKey = ExchangeKey(
+            keyType,
+            null,
+            null,
+            key.getKeyBase64(),
+            key.keyVersion
+        )
+        val json = moshi.adapter(ExchangeKey::class.java).toJson(exchangeKey)
+        secureStore.addData(alias, json.toCharArray())
+    }
+
+    override fun storeKey(alias: String, key: GCKeyPair) {
+        val exchangeKey = ExchangeKey(
+            KeyType.APP_PRIVATE_KEY,
+            key.getPrivateKeyBase64(),
+            key.getPublicKeyBase64(),
+            null,
+            key.keyVersion
+        )
+        val json = moshi.adapter(ExchangeKey::class.java).toJson(exchangeKey)
+        secureStore.addData(alias, json.toCharArray())
+    }
+
+    @Throws(IOException::class)
+    override fun getExchangeKey(alias: String): ExchangeKey {
+        val data = secureStore.getData(alias) ?: CharArray(0)
```

Comment:
Why is this returning an empty array in case the store doesn't contain data?

Reply:
Good catch! The original would simply passthrough the return value, which can be null.`fromJson` does not allow a null as parameter. But if we delegate an empty CharArray the old behaviour is preserved (sort of).
However nicer would be to fail in here then in moshi, but this is a change when can postpone until we touch the tests.

Reply:
hmm, null was a testable error case. What about?
```
    @Throws(IOException::class)
    fun getExchangeKey(alias: String): ExchangeKey {
        val json = secureStore.getData(alias)
            ?: throw (DecryptionFailed("Failed to decrypt data"))

        return moshi.adapter(ExchangeKey::class.java).fromJson(json.toString())
            ?: throw (DecryptionFailed("Failed to decrypt data"))
    }
```

Reply:
Done